src/odemis/acq/acqmng.py

```diff
@@ -83,6 +85,205 @@ def acquire(streams, settings_obs=None):
+                    if self._future_state == CANCELLED:
+                        raise CancelledError()
+
+                    # check on the acquired data
+                    if not data:
+                        logging.warning("The acquired data array for stream %s is empty", stream)
+                    else:
+                        logging.info("The acquisition for stream %s is done", stream)
+
+                except CancelledError:
+                    raise CancelledError()
+                except Exception as e:
+                    logging.exception("The acquisition failed because of %s" %e)
+                    # return what was acquired so far  
+                    return acquired_data, e
+                finally:
+                    acquired_data.append(data[0])
+                    # update the remaining time
+                    remaining_t -= stream.estimateAcquisitionTime()
+                    self._main_future.set_end_time(time.time() + remaining_t)
+
+            else: # stream can be zstack imaged 
+                # for each stream, iterate through zlevels (if exist)
+                for i, z in enumerate(self._zpos):
+                    # move the focuser
+                    self._actuator_f = stream.focuser.moveAbs({"z": z})
+                    self._actuator_f.result()
+
+                    # check if cancellation happened while the actuator future is working
+                    if self._future_state == CANCELLED:
+                        raise CancelledError()
+
+                    # subtract one zstep time
+                    if i != len(self._zpos) - 1:
+                        remaining_t -= self._zstep_duration
+                        self._main_future.set_end_time(time.time() + remaining_t)
+
+                    try:
+                        # acquire this single stream, and get the data
+                        self._single_acqui_f = acquire([stream], self._settings_obs)
+                        data, exp = self._single_acqui_f.result()
+
+                        # check if cancellation happened while the acquiring future is working
+                        if self._future_state == CANCELLED:
+                            raise CancelledError()
+
+                        # check on the acquired data
+                        if not data:
+                            logging.warning("The acquired data array for stream %s is empty", stream)
+                        else:
+                            logging.info("The acquisition for stream %s is done", stream)
+
+                    except CancelledError:
+                        raise CancelledError()
+                    except Exception as exp:
+                        logging.exception("The acquisition failed at the %s-th zlevel of streams %s because of %s" %(i+1, stream, exp))
+                        # assemble what was acquired so far, and return 
+                        zcube = assembleZCube(zstack, self._zpos)
```

Comment:
That probably won't work as it probably failed before the very last step. So zlevels (zpos here) will not be the same length as ztack. So you'd need to pass a shorten zlevels. Also you need to handle the case of zstack being empty (not sure if assembleZCube already handles this).
An alternative  would be to not support the ztack reconstruction in case of error for now (and instead just put a TODO).